src/otaclient/boot_control/_grub.py

```diff
@@ -763,36 +763,53 @@ def _update_fstab(self, *, active_slot_fstab: Path, standby_slot_fstab: Path):
+            """Create a dictionary mapping mount points to regex match objects from fstab"""
+            content = read_str_from_file(fstab_path)
+            return {
+                m.group("mount_point"): m
+                for m in (fstab_entry_pa.match(line) for line in content.splitlines())
+                if m
+            }
 
-        for line in fstab_standby.splitlines():
-            ma = fstab_entry_pa.match(line)
-            if ma and ma.group("mount_point") != "/":
-                fstab_standby_dict[ma.group("mount_point")] = ma
+        standby_dict = read_fstab_dict(standby_slot_fstab)
+        active_dict = read_fstab_dict(active_slot_fstab)
 
-        # merge entries
         merged: List[str] = []
-        fstab_active = read_str_from_file(active_slot_fstab)
-        for line in fstab_active.splitlines():
-            if ma := fstab_entry_pa.match(line):
-                mp = ma.group("mount_point")
-                if mp == "/":  # rootfs mp, unconditionally replace uuid
-                    _list = list(ma.groups())
-                    _list[0] = standby_uuid_str
-                    merged.append("\t".join(_list))
-                elif mp in fstab_standby_dict:
-                    merged.append("\t".join(fstab_standby_dict[mp].groups()))
-                    del fstab_standby_dict[mp]
+        found_root = False
+
+        # Merge using standby as the base
+        for line in read_str_from_file(standby_slot_fstab).splitlines():
+            if standby_entry_match := fstab_entry_pa.match(line):
+                mount_point = standby_entry_match.group("mount_point")
+                if mount_point == cfg.CANONICAL_ROOT:
+                    # Use active "/" entry as base, but replace UUID with standby_uuid
+                    active_ma = active_dict[mount_point]
+                    merged.append(
+                        "\t".join([standby_uuid_str] + list(active_ma.groups())[1:])
+                    )
+                    found_root = True
+                elif mount_point in (cfg.BOOT_DPATH, cfg.EFI_DPATH):
+                    # base mount points(/boot, /boot/efi) should be merged from active fstab
+                    # These base mount points are created by USB Installer and not in project settings,
+                    # so we need to preserve them from active slot's fstab.
+                    # Reference: https://tier4.atlassian.net/browse/T4DEV-39187
+                    entry = active_dict.get(mount_point, standby_entry_match)
+                    merged.append("\t".join(entry.groups()))
                 else:
-                    merged.append("\t".join(ma.groups()))
-            elif line.strip().startswith("#"):
-                merged.append(line)  # re-add comments to merged
+                    merged.append("\t".join(standby_entry_match.groups()))
+            else:
+                merged.append(line)  # Keep comments and blank lines
```

Comment:
After some considerations, I think let's just only add valid mount point entry into the fstab. 
`update_fstab` method should be able to correct the invalid fstab in some extend(like filtering away invalid mount point entries).

Also, comments are not so necessary, besides `/`, `/boot` and `/boot/efi`, all other entries must either come from pilot-auto integration, or from post OTA services, we can track down the origin without relying on comments.

Reply:
The new implement is much cleaner and more comprehensive than the aged old implementation, good job!